### PR TITLE
Fix Dockerfile; add YAML for decentralized devstack settings

### PR DESCRIPTION
**Pull request by kdmccormick**
TLDR: These changes are part of Decentralized Devstack(DD) work and will help us with prototyping DD with enterprise-catalog.

For context, see [spike notes](https://openedx.atlassian.net/wiki/spaces/ENG/pages/1653702661/Devstack+2.0+Spike+Notes) and [post-spike action plan](https://openedx.atlassian.net/wiki/spaces/ENG/pages/1685619678/Post-Spike+Action+Plan).

Platform PR: https://github.com/edx/edx-platform/pull/24608
Enterprise Catalog PR: https://github.com/edx/enterprise-catalog/pull/156

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -1,35 +1,66 @@
...
+COPY requirements/ requirements/
+RUN pip install -r requirements/production.txt
+
+# Copy over rest of code.
+# We do this AFTER requirements so that the requirements cache isn't busted
+# every time any bit of code is changed.
+COPY . .
+
+# Expose canoncal Discovery port (TODO: This should be 8381 for prod...)
+EXPOSE 18381
```
Comment (kdmccormick) [inline Dockerfile]: Need to make this 8381 for prod, and 18381 for dev.

File: Dockerfile [hunk 2]
```diff
@@ -1,35 +1,66 @@
...
 ENV LANGUAGE en_US:en
 ENV LC_ALL en_US.UTF-8
 
-COPY . /edx/app/discovery
-WORKDIR /edx/app/discovery
+# Make necessary directories and environment variables.
+RUN mkdir -p /edx/var/discovery/staticfiles
+RUN mkdir -p /edx/var/discovery/media
+ENV DJANGO_SETTINGS_MODULE course_discovery.settings.production
+ENV DISCOVERY_CFG /edx/app/discovery/devstack.yml
```
Comment (kdmccormick) [inline Dockerfile]: (This is one of the action items in the [post-spike action plan](https://openedx.atlassian.net/wiki/spaces/ENG/pages/1685619678/Post-Spike+Action+Plan?focusedCommentId=1746436117#comment-1746436117) under **Reconcile settings handling between edx-platform and course-discovery PRs.**).

In this PR, we use production settings but pull in a YAML file with devstack-like settings.

In [the edx-platform PR](https://github.com/edx/edx-platform/pull/24608/files#diff-3254677a7917c6c01f55212f86c57fbfR119) , though, we pull in a new settings file `devstack_decentralized` and use a YAML file copied in from decentralized devstack.

It'd be good to standardize on one of those two approaches before merging.

File: Dockerfile [hunk 3]
```diff
@@ -1,35 +1,66 @@
...
 
-RUN apt update && \
-  apt install -y git-core language-pack-en python3.5 python3-pip python3.5-dev \
-  build-essential libffi-dev libmysqlclient-dev libxml2-dev libxslt-dev libjpeg-dev libssl-dev && \
-  pip3 install nodeenv && \
-  pip3 install --upgrade pip setuptools && \
-  rm -rf /var/lib/apt/lists/*
+# System requirements.
+RUN apt-get update
+RUN apt-get install -y \
```
Comment (kdmccormick) [inline Dockerfile]:
```suggestion
RUN apt-get install --yes \
```

File: Dockerfile [hunk 4]
```diff
@@ -1,35 +1,66 @@
...
+ENV DISCOVERY_CFG /edx/app/discovery/devstack.yml
 
-ENV DISCOVERY_CFG /edx/etc/discovery.yml
+# Working directory will be root of repo.
+WORKDIR /edx/app/discovery
 
+# Copy just JS requirements and install them.
+COPY package.json package.json
+COPY package-lock.json package-lock.json
 RUN nodeenv /edx/app/nodeenv --node=8.9.3 --prebuilt
```
Comment (kdmccormick) [inline Dockerfile]: Aren't we on Node 12?

File: Dockerfile [hunk 5]
```diff
@@ -1,35 +1,66 @@
...
 ENV LANGUAGE en_US:en
 ENV LC_ALL en_US.UTF-8
 
-COPY . /edx/app/discovery
-WORKDIR /edx/app/discovery
+# Make necessary directories and environment variables.
+RUN mkdir -p /edx/var/discovery/staticfiles
+RUN mkdir -p /edx/var/discovery/media
+ENV DJANGO_SETTINGS_MODULE course_discovery.settings.production
+ENV DISCOVERY_CFG /edx/app/discovery/devstack.yml
```
Comment (kdmccormick) [inline Dockerfile:36]: We should build static assets in this image.

File: Dockerfile [hunk 6]
```diff
@@ -1,35 +1,66 @@
 FROM ubuntu:xenial as app
```
Comment (kdmccormick) [inline Dockerfile:2]: @adzuci Do you think we should add a GitHub action to this PR to push the images?

Comment (kdmccormick) [reply Dockerfile:2]: On second thought, in the interest of doing work in parallel, let's make that its own PR.

Comment (adzuci) [reply Dockerfile:2]: I updated the confluence page with some more information.  This image is being build and pushed to https://hub.docker.com/repository/docker/openedx/course-discovery currently, but the build stages and tags may need to be updated.

Comment (kdmccormick) [reply Dockerfile:2]: @adzuci Cool; I forgot that work had been done! Thanks.

File: Dockerfile (hunk 2 above)
Comment (jinder1s) [reply Dockerfile]: In spirit of getting to prototype quickly, for now, we plan on leaving this as is and clean this back up in next epic.  [ARCHBOM-1445](https://openedx.atlassian.net/browse/ARCHBOM-1445) will pick this work up in a different PR.

File: Dockerfile (hunk 5 above)
Comment (jinder1s) [reply Dockerfile:36]: This work was done in: #2718

**Review by jinder1s — APPROVED** [review]

**Review by Dillon-Dumesnil — APPROVED** [review]
Sure, this all looks fine. One thing I did want to call out just so we can keep it in mind is that if we are updating all services and removing the 10000 prefix to the ports, we will also need to update the `CORS_ORIGIN_WHITELIST` in these devstack.yml files to also adjust for that. (Thinking of frontend-app-publisher `18400` and frontend-app-support-tools `18450` in this particular devstack.yml)

**Pull request merged**